apis/runtime/v1alpha1/api.proto

```diff
@@ -0,0 +1,158 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// To regenerate api.pb.go run scripts/generate_runtime.sh
+syntax = "proto3";
+
+package runtime.v1alpha1;
+option go_package = "github.com/koordinator-sh/koordinator/apis/runtime/v1alpha1";
+
+
+// PodSandboxMetadata holds all necessary information for sandbox name.
+message PodSandboxMetadata {
+  // Pod name of the sandbox. Same as the pod name in the Pod ObjectMeta.
+  string name = 1;
+  // Pod UID of the sandbox. Same as the pod UID in the Pod ObjectMeta.
+  string uid = 2;
+  // Pod namespace of the sandbox. Same as the pod namespace in the Pod ObjectMeta.
+  string namespace = 3;
+  // Attempt number of creating the sandbox. Default: 0.
+  uint32 attempt = 4;
+}
+
+// PreRunPodSandboxHookRequest is sent to RuntimeHookServer before pod creating request transferred to
+// backend containerd or dockerd. This Request is generated basing on CRI's RunPodSandboxRequest, including pod
+// Meta, Resources, Annotations .e.g and RuntimeHookServer should ensure correct operations basing on
+// this request.
+message PreRunPodSandboxHookRequest {
+  // Metadata of the sandbox. This information will uniquely identify the sandbox.
+  PodSandboxMetadata pod_meta = 1;
+  // Named runtime to use for podSandbox.
+  string runtime_handler = 2;
+  // Labels/Annotations are key-value pairs that may be used.
+  map<string, string> labels = 3;
+  map<string, string> annotations = 4;
+  // Parent cgroup of the pod sandbox.
+  string cgroup_parent = 5;
+  // Optional overhead represents the overheads associated with this sandbox,
+  // same as LinuxPodSandboxConfig.overhead under CRI scenario
+  LinuxContainerResources overhead = 6;
+  // Optional resources represents the sum of container resources for this sandbox
+  // same as LinuxPodSandboxConfig.resources under CRI scenario
+  LinuxContainerResources resources = 7;
+}
+
+// PreRunPodSandboxHookResponse is RuntimeHookServer's response to PreRunPodSandboxHookRequest.
+// RuntimeManager will merge PreRunPodSandboxHookResponse and RunPodSandboxRequest to generate a
+// RunPodSandboxRequest to containerd(dockerd).
+message PreRunPodSandboxHookResponse{
+  // RuntimeHookServer may inject additional labels/annotations to PodSandboxRequest to be send to
+  // backend contaienrd/dockerd
+  map<string, string> labels = 1;
+  map<string, string> annotations = 2;
+  // RuntimeHookSever may modify cgroup_parent to construct customized cgroup topology.
+  string cgroup_parent = 3;
+  // RuntimeHookServer may modify the linux resource config.
```

Comment:
overhead&resources wont be modified,

Reply:
resources info would be assigned to pod-level cgroup.
so it makes sense to allow HookServert to change pod-level cgroup config here.

overhead can be removed.

Reply:
> overhead&resources wont be modified, so remove those fields from response ?

resources info would be assigned to pod-level cgroup.
so it makes sense to allow HookServert to change pod-level cgroup config here.

overhead can be removed.

Reply:
when do you need to change this resource, to make response.resources different than request.resources?

Reply:
> when do you need to change this resource, to make response.resources different than request.resources?

for batch pod,  lower cpu preemption priority by setting pod's pod-level cgroup cpu.shares to 2